Source/DafnyCompilers/CompilerRewriter.dfy

```diff
@@ -0,0 +1,2215 @@
+        // We need the argument types to be smaller than the closure types, to prove termination.\
+        // In effect, the arguments types should be given by the closure's input types.
+        && (forall i | 0 <= i < |vars| :: ValueTypeHeight(argvs[i]) < ValueTypeHeight(v))
+        && (forall i | 0 <= i < |vars| :: ValueTypeHeight(argvs'[i]) < ValueTypeHeight(v'))
+        && (forall i | 0 <= i < |vars| :: EqValue(argvs[i], argvs'[i])) ::
+        var res := InterpCallFunctionBody(v, fuel, argvs);
+        var res' := InterpCallFunctionBody(v', fuel, argvs');
+        // We can't use naked functions in recursive setting: this forces us to write the expanded
+        // match rather than using an auxiliary function like `EqPureInterpResult`.
+        match (res, res') {
+          case (Success(ov), Success(ov')) =>
+            // We need to assume those assertions to prove termination: the value returned by a closure
+            // has a type which is smaller than the closure type (its type is given by the closure return
+            // type)
+            assume ValueTypeHeight(ov) < ValueTypeHeight(v);
+            assume ValueTypeHeight(ov') < ValueTypeHeight(v');
+            EqValue(ov, ov')
+          case (Failure(_), Failure(_)) =>
+            true
+          case _ =>
+            false
+        })
+    }
+
+    predicate {:verify true} EqState(ctx: State, ctx': State)
+    {
+      GEqState(EqValue, ctx, ctx')
+    }
+
+    predicate {:verify true} EqCtx(ctx: Context, ctx': Context)
+      requires WellFormedContext(ctx)
+      requires WellFormedContext(ctx')
+    {
+      GEqCtx(EqValue, ctx, ctx')
+    }
+
+    predicate {:verify true} EqInterpResult<T(0)>(
+      eq_value: (T,T) -> bool, res: InterpResult<T>, res': InterpResult<T>)
+    {
+      GEqInterpResult(Mk_EqState(EqValue), eq_value, res, res')
+    }
+
+    lemma {:verify true} EqValueHasEq_Lem(v: WV, v': WV)
+      requires EqValue(v,v')
+      requires HasEqValue(v)
+      requires HasEqValue(v')
+      ensures v == v'
+    // If values are equivalent and have a decidable equality, they are necessarily equal.
+    {
+      reveal EqValue_Closure();
+    }
+
+    lemma {:verify true} EqValueHasEq_Forall_Lem()
+      ensures forall v: WV, v': WV | EqValue(v,v') && HasEqValue(v) && HasEqValue(v') :: v == v'
+    {
+      forall v: WV, v': WV | EqValue(v,v') && HasEqValue(v) && HasEqValue(v') ensures v == v' {
+        EqValueHasEq_Lem(v, v');
+      }
+    }
+
+    // TODO: move
+    lemma {:verify true} EqInterp_Expr_EqState_Lem(e: Exprs.T, fuel: nat, ctx: State, ctx': State)
+      requires SupportsInterp(e)
+      requires EqState(ctx, ctx')
+      ensures EqInterpResultValue(InterpExpr(e, fuel, ctx), InterpExpr(e, fuel, ctx'))
+    // The proof should be similar to ``EqInterp_Expr_CanBeMapLifted_Lem`` (and actually
+    // simpler), but I'm not sure how to efficiently factorize the two.
+    {
```

Comment:
You can omit the body of a lemma instead of putting an assume false in it
(Although arguably this way is more explicit.)